samtranslator/plugins/api/implicit_api_plugin.py

```diff
@@ -247,6 +257,89 @@ def _maybe_add_condition_to_implicit_api(self, template_dict):
                     template_dict, self.implicit_api_condition, all_resource_method_conditions
                 )
 
+    def _maybe_add_deletion_policy_to_implicit_api(self, template_dict):
+        """
+        Decides whether to add a deletion policy to the implicit api resource.
+        :param dict template_dict: SAM template dictionary
+        """
+        # Short-circuit if template doesn't have any functions with implicit API events
+        if not self.api_deletion_policies.get(self.implicit_api_logical_id, {}):
+            return
+
+        # Add a deletion policy to the API resource if its resources contains DeletionPolicy.
+        implicit_api_deletion_policies = self.api_deletion_policies[self.implicit_api_logical_id]
+        all_resource_method_deletion_policies = set(
+            [
+                deletion_policy
+                for path, method_deletion_policies in implicit_api_deletion_policies.items()
+                for method, deletion_policy in method_deletion_policies.items()
+            ]
+        )
+        at_least_one_resource_method = len(all_resource_method_deletion_policies) > 0
+        one_resource_method_contains_deletion_policy = False
+        contains_retain = False
+        contains_delete = False
+        # If multiple functions with multiple different policies reference the Implicit Api,
+        # we set DeletionPolicy to Retain if Retain is present in one of the functions,
+        # else Delete if Delete is present
+        for iterated_policy in all_resource_method_deletion_policies:
+            if iterated_policy:
+                one_resource_method_contains_deletion_policy = True
+                if iterated_policy == "Retain":
+                    contains_retain = True
+                if iterated_policy == "Delete":
+                    contains_delete = True
+        if at_least_one_resource_method and one_resource_method_contains_deletion_policy:
+            implicit_api_resource = template_dict.get("Resources").get(self.implicit_api_logical_id)
+            if contains_retain:
+                implicit_api_resource["DeletionPolicy"] = "Retain"
```

Comment:
Is `implicit_api_resource` proved to be a dict?

Reply:
Yes it is